### PR TITLE
blockdev_mirror_no_space: Mirror to an image without enough space

### PR DESCRIPTION
**Pull request by zhencliu**
Signed-off-by: Zhenchao Liu <zhencliu@redhat.com>

ID: 1926272

### CODE REVIEW

**Comment by zhencliu** [comment]
local fs:
 (1/1) Host_RHEL.m8.u3.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.blockdev_mirror_no_space.q35: PASS (135.84 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2021-02-08T06.09-092fba0/results.html
JOB TIME   : 137.58 s

nbd:
 (1/1) Host_RHEL.m8.u3.nbd.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.blockdev_mirror_no_space: PASS (144.25 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2021-02-08T06.52-6e52cbb/results.html
JOB TIME   : 145.98 s

iscsi direct:
 (3/3) Host_RHEL.m8.u3.iscsi_direct.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.blockdev_mirror_no_space.q35: PASS (154.25 s)

ceph:
 (1/1) Host_RHEL.m8.u3.ceph.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_no_space.q35: PASS (191.85 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2021-02-08T07.02-500dc59/results.html
JOB TIME   : 193.63 s

**Comment by zhencliu** [comment]
hi @ngu-niny @YongxueHong , can you please help to review the patch? Thanks.

**Comment by ngu-niny** [comment]
@zhencliu Test PASS, but it seems there are 3 times vm start/quit in the case, and the block-morror job is started until the 3rd time to start the vm 

(05/14) Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space:  PASS (304.85 s)

 (12/14) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space:  PASS (359.05 s)

**Comment by zhencliu** [comment]
> @zhencliu Test PASS, but it seems there are 3 times vm start/quit in the case, and the block-mirror job is started until the 3rd time booting the vm.
> 
> (05/14) Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space: PASS (304.85 s)
> 
> (12/14) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space: PASS (359.05 s)

Great catch, thank you Nini. That's because I let VT  start VM automatically, but in the base class, we stop VM if it's alive and then restart it. So you could see several VM startup/stop, I just set start_vm = no, then let test case start VM by itself, which can avoid the issue.

BTW: Even though VM was started several times, you can see that the testing strategy/assertion is correct, the issue is it costs too much time.

Thanks much. - Zhenchao

**Comment by zhencliu** [comment]
(1/1) Host_RHEL.m9.u0.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_no_space.q35: PASS (116.86 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-02-10T06.38-461a64e/results.html
JOB TIME   : 117.73 s

**Comment by ngu-niny** [comment]
Test PASS:

'1-Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space' Finshed 2021-02-20 09:34:56 2021-02-20 09:36:59 PASS 123.473426
'2-Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space' Finshed 2021-02-20 09:37:00 2021-02-20 09:39:34 PASS 154.3696

'1-Host_RHEL.m8.u4.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space' Finshed 2021-02-20 09:43:26 2021-02-20 09:45:53 PASS 147.72551
'2-Host_RHEL.m8.u4.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_no_space' Finshed 2021-02-20 09:45:53 2021-02-20 09:48:02 PASS 128.651397

Acked-by: Nini Gu <ngu@redhat.com>

File: qemu/tests/cfg/blockdev_mirror_no_space.cfg [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    backup_options_image1 = sync
+    sync = full
+    storage_pools = default
+    storage_pool = default
+    error_msg = 'No space left on device'
+
+    # Create mirror1 image under target_path
+    target_path = /tmp/tmp_target_path
+    tmp_image_file = /tmp/tmp_image_file
+    pre_command = "mkdir -p ${target_path} && dd if=/dev/urandom of=${tmp_image_file} bs=1M count=100 && mkfs.ext4 ${tmp_image_file} && mount -o loop ${tmp_image_file} ${target_path}"
```
Comment (YongxueHong) [inline qemu/tests/cfg/blockdev_mirror_no_space.cfg:28]:
```suggestion
    pre_command = "mkdir -p ${target_path} && dd if=/dev/urandom of=${tmp_image_file} bs=1M count=100 &&"
    pre_command += " mkfs.ext4 ${tmp_image_file} && mount -o loop ${tmp_image_file} ${target_path}"
```
a little longer :)

**Review by YongxueHong — APPROVED** [review]
LGTM.

File: qemu/tests/cfg/blockdev_mirror_no_space.cfg (hunk 1 above)
Comment (zhencliu) [reply qemu/tests/cfg/blockdev_mirror_no_space.cfg:28]: OK, will update it later

**Comment by zhencliu** [comment]
(1/1) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.blockdev_mirror_no_space.q35: PASS (81.85 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2021-02-23T23.24-9820c20/results.html
JOB TIME   : 83.92 s

Refer to @YongxueHong 's comment, split one long line in cfg file into 2 short lines, test passed.
Merge it.

**Pull request merged**